node/workflow-api.md

```diff
@@ -0,0 +1,322 @@
+    // Main workflow logic
+    async execute() {
+      await unblocked;
+      isBlocked = false;
+    },
+    signals: {
+      unblock() {
+        unblocked.resolve();
+      },
+    },
+    queries: {
+      isBlocked() {
+        return isBlocked;
+      },
+    },
+  };
+};
+```
+
+#### Issues
+
+- Types must be explicitly declared (To make [eslint happy](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md))
+- Very verbose
+- Trigger (ResolvablePromise) is a foreign concept to JS developers
+- Trigger doesn't provide all of the functionality for this workflow, an additional `isBlocked` boolean variable is required
+- Signals and queries must be defined at the top level in order for types to be inferred.
+
+#### Advantages
+
+- Types for all Workflow handlers can be inferred at Workflow handle creation
+
+### Goals for the redesign
+
+- Maintain type inference capabilities - Mandatory for the Typescript SDK
+- Dynamic signals and queries registration first - A single API for Workflow handler registration
+- Componentization of related functionality - increase reusability
+- Implicit interface declarations - less boilerplate
+- No global variables - to support running Workflows in less isolated runtimes
+
+### New proposal
+
+#### Remove the `Trigger` API in favor of `condition`
+
+`condition` is similar to `Workflow.await` in Java, it returns a Promise and accepts a closure which resolve the Promise when it returns `true`.
+
+```ts
+import { condition } from '@temporalio/workflow';
+
+let isBlocked = true;
+await condition(() => !isBlocked);
+```
+
+#### Remove the Workflow factory wrapper and always use dynamic handler registration
+
+`workflows.ts`
+
+```ts
+import { Signal, Query, condition } from '@temporalio/workflow';
+
+// Define the signal / query with the type signature at the top level to share the definition for handle creation
+export const unblockSignal = new Signal<[] /* args */>('unblock');
+export const isBlockedQuery = new Query<[] /* args */, boolean /* return type */>('isBlocked');
+
+// Types are inferred without an additional type definition
+export async function myWorkflow(arg1: number, arg2: string): Promise<void> {
+  let isBlocked = true;
+  unblockSignal.handler = () => (isBlocked = false);
+  isBlockedQuery.handler = () => isBlocked;
+  await condition(() => !isBlocked);
+}
+```
+
+- Note how all unblocking related functionality is co-located now, we can use this to split our Workflow into components, for example we could have a reusable "unblock" component.
+
+  ```ts
+  export async function unblocked() {
+    let isBlocked = true;
+    unblockSignal.handler = () => (isBlocked = false);
+    isBlockedQuery.handler = () => isBlocked;
+    await condition(() => !isBlocked);
+  }
+
+  export async function myWorkflow1() {
+    await unblocked();
+  }
+
+  export async function myWorkflow2() {
+    await unblocked();
+  }
```

Comment:
Maxim concern: stealing signals from each other like in Go

Reply:
They are different signals. You are reusing signal handling code, not signals. At the end of the day every signal is uniquely bound to a workflow run (consider: when you send signals they go a specific run).

So anything that would allow these two awaits to be anything other than completely independent is a bug. A user could maybe do something nasty that we can't really prevent like make a program wide static value that the shared signal code mutates, in which case they are now tied together, but that's a big no-no (and is something they could've done from just raw workflow code anyway).

What this does say to me is that perhaps we should call 
```
// Define the signal / query with the type signature at the top level to share the definition for handle creation
export const unblockSignal = new Signal<[] /* args */>('unblock');
export const isBlockedQuery = new Query<[] /* args */, boolean /* return type */>('isBlocked');
```

`SignalSignature` and `QuerySignature` (or `Definition` or `Type`). Because when you define them this way that's sort of more accurately what you're doing.

Even more accurate would be to only allow attaching the signal's *identifier* inside of the workflow. In this case you would have something like:

```
export const unblockSignalDef = new SignalDefinition<[]>();

export async function unblocked() {
  let isBlocked = true;
  unblockSignalDef.handler('unblock') = () => (isBlocked = false);
  await condition(() => !isBlocked);
}
```


Both ways have merit. What I demonstrate I think maybe makes the concepts a bit more clear and makes it easy to re-use the same input/return types for multiple signals with different identifiers which is maybe desirable

Reply:
Yes, these signals are instantiated per WF.
I also suggested to @sw-yx at some point that the API should be called `defineSignal` to explain that this is just a definition.
Worth thinking about the naming a bit more.

In any case the signal definition should include the signal name for it to be used by both WF and client.

Reply:
@bergundy The new proposal reminds me of EventEmitter, which at first glance seems appropriate.

However, I do think `on` is too concise. _Too cute_ is how I would put it. It makes me wonder whether workflow is an EventEmitter somehow. (Spoiler Alert: It is not.)

I would choose a more descriptive name and let the user alias if desired:

```
import { defineSignal, defineQuery, setListener as on } from '@temporalio/workflow';

// ...
on(unblockSignal, () => (isBlocked = false));
```

On second thought, though, it bothers me that signal and query are conflated in this way.

Fundamentally, I think of **signal** as something that may have a _list of listeners_, like an EventEmitter. The use of `on` or `add` prefixes reinforces this idea -- once the EventEmitter analogy is activated.

Whereas a **query** returns a result, hence one body.

Bottom line: I'd like to see more descriptive names that reinforce the differences between signals and queries, and let the user conflate them if desired.

Reply:
Having this as a function defined on the workflow module is great. Totally clears up the things I mentioned. Like that a lot.

As for naming, I'm of two minds. The conciseness of `on` is a nice properly IMO. The type system will be able to specialize the second argument based on the first (IE: Signals return void, queries return something), and I think that is pretty much sufficient signal to the user.

OTOH, there's nothing wrong with `onSignal` and `onQuery` to differentiate. Queries *cannot* mutate state, and unfortunately we can't really enforce that, so making that clear in the docstring for `onQuery` is a nice property.

RE: Multiple handlers. I can see why it makes sense, especially in the "JS event handler" context. I think I'd prefer us to keep the underlying primitives as simple as possible and let the user layer on top of that. If the need that kind of multiple dispatch behavior, they can easily implement a top-level handler that calls a bunch of other stuff. My main reason for that is you probably end up with users asking for stuff like "I need a function that clears the event handler list", etc, other stuff related to managing that implicit set of handlers. If lots of users want the dispatch behavior, we can always layer it on by composition (ex: `LayerableSignalDefinition` or `onSignalAdditional` or whatever) rather than baking it in to the base primitive.